bst.py

```diff
@@ -0,0 +1,179 @@
+        Return 0 if tree is empty."""
+        if self.root is not None:
+            l, r = self._lr_levels()
+            return max([l, r]) + 1
+        else:
+            return 0
+
+    def balance(self):
+        """Return the difference between the number of nodes on the left with
+        respect to the right side.
+        --Negative if more on right
+        --Positive if more on left
+        --Zero if tree is perfectly balanced"""
+        if self.root is not None:
+            l, r = self._lr_levels()
+            return l - r
+        else:
+            return 0
+
+    def _find(self, val):
+        """Return a tuple containing (node, side) with target val if it exists,
+        otherwise return the would be parent and 1 if it is on the left, -1 if
+        it is on the right, 0 if it is the root."""
+
+        def _look(node):
+            if node.val > val:
+                if node.left is not None:
+                    return _look(node.left)
+            elif node.val < val:
+                if node.right is not None:
+                    return _look(node.right)
+            return node
+
+        if self.root is not None:
+            return _look(self.root)
+        else:
+            return None
+
+    def _lr_levels(self):
+        def _levels(node):
+            l, r = 0, 0
+            if node.left is not None:
+                l = _levels(node.left)
+            if node.right is not None:
+                r = _levels(node.right)
+            return max([l, r]) + 1
+
+        if self.root is None:
+            return None
+        w_left, w_right = 0, 0
+        if self.root.left is not None:
+            w_left = _levels(self.root.left)
+        if self.root.right is not None:
+            w_right = _levels(self.root.right)
+        return w_left, w_right
+
+
+if __name__ == '__main__':
+    # Worst case senerio, the BST is completely unbalanced:
+    # either:
+    # --the root is the smallest item and you are searching
+    # for an item bigger than the biggest item in the BST, or
+    # --the root is the biggest item and you are searching
+    # for an item smaller than the smallest item in the list.
+    BST1 = BST()
+    for x in range(51):
+        BST1.insert(x)
+    start = default_timer()
+    for x in range(10000):
+        BST1.contains(51)
+    print BST1.contains(51)
+    end = default_timer()
+    print str(end - start) + "<-- Quite a while!"
+    # Best case senario, the item you are looking for is the item at root.
+    BST2 = BST(1)
+    start = default_timer()
+    for x in range(10000):
+        BST2.contains(1)
+    print BST2.contains(1)
+    end = default_timer()
+    print str(end - start) + "<-- Quite fast!"
+    # Or that the tree is well balanced.
+    BST3 = BST()
+    l = [50, 25, 75, 12, 37, 63, 87, 6, 18, 31, 43, 57, 69, 81, 93, 3, 9, 15,
+         21, 28, 34, 40, 46, 54, 60, 66, 72, 78, 84, 90, 96, 1, 4, 8, 10, 14,
+         6, 20, 22, 27, 29, 39, 41, 59, 61, 65, 67, 77, 79, 89, 91]
```

Comment:
Can you come up with a way to manufacture this list without hand-typing it?  Given the nature of a BST, it should be possible to do so.